vllm/core/scheduler.py

```diff
@@ -364,10 +364,35 @@ def __init__(
         self.num_cumulative_preemption: int = 0
 
         # Used to cache python objects
-        self._scheduler_running_outputs_cache: PyObjectCache = PyObjectCache(
-            scheduler_running_outputs_builder)
-        self._scheduled_seq_group_cache: PyObjectCache = PyObjectCache(
-            scheduled_seq_group_builder)
+        self._seq_group_metadata_cache: List[PyObjectCache] = []
+        self._scheduler_running_outputs_cache: List[PyObjectCache] = []
+        self._scheduled_seq_group_cache: List[PyObjectCache] = []
+
+        # For async output processing, we need to swap cache buffers between
+        # iterations. I.e. since the output processing is lagged one step,
+        # we cannot reuse the cached objects immediately when the schedule()
+        # is called again, but only when schedule() is called the second time.
+        self.output_proc_callback_fn = output_proc_callback_fn
+        self.use_output_proc_callback = self.output_proc_callback_fn is not None
+        self.num_cache_iters = 2 if self.use_output_proc_callback else 1
+
+        self.cache_id = 0
+        for i in range(self.num_cache_iters):
+            self._seq_group_metadata_cache.append(
+                PyObjectCache(seq_group_metadata_builder))
+            self._scheduler_running_outputs_cache.append(
+                PyObjectCache(scheduler_running_outputs_builder))
+            self._scheduled_seq_group_cache.append(
+                PyObjectCache(scheduled_seq_group_builder))
+
+        # Avoid deque alloc
+        self.tmp_queue: Deque[SequenceGroup] = deque()
```

Comment:
Can you say more about this? Or use a better name for this queue. It's hard to catch its functionality in the rest of its use cases.

Reply:
Its a deque acting as a temporary buffer to store the new seqs in running queue. It's only used inside `free_finished_seq_groups`, we save some time by avoiding re-allocation during each step.